cmd/core/pdr.go

```diff
@@ -0,0 +1,141 @@
+package core
+
+import (
+	"fmt"
+	"github.com/edgecomllc/eupf/cmd/core/service"
+	"github.com/edgecomllc/eupf/cmd/ebpf"
+	"github.com/rs/zerolog/log"
+	"github.com/wmnsk/go-pfcp/ie"
+	"net"
+)
+
+func deletePDR(spdrInfo SPDRInfo, mapOperations ebpf.ForwardingPlaneController, resourceManager *service.ResourceManager, seID uint64) error {
+	if spdrInfo.Ipv4 != nil {
+		if err := mapOperations.DeletePdrDownLink(spdrInfo.Ipv4); err != nil {
+			return fmt.Errorf("Can't delete IPv4 PDR: %s", err.Error())
+		}
+	} else if spdrInfo.Ipv6 != nil {
+		if err := mapOperations.DeleteDownlinkPdrIp6(spdrInfo.Ipv6); err != nil {
+			return fmt.Errorf("Can't delete IPv6 PDR: %s", err.Error())
+		}
+	} else {
+		if err := mapOperations.DeletePdrUpLink(spdrInfo.Teid); err != nil {
+			return fmt.Errorf("Can't delete GTP PDR: %s", err.Error())
+		}
+	}
+	if spdrInfo.Teid != 0 {
+		if resourceManager.FTEIDM != nil {
+			resourceManager.FTEIDM.AllocateTEID(seID, spdrInfo.PdrID)
+		}
+	}
+	return nil
+}
+
+// func extractPDR(pdr *ie.IE, session *Session, spdrInfo *SPDRInfo, resourceManager *service.ResourceManager, teidCache map[uint8]uint32) error {
+func extractPDR(pdr *ie.IE, spdrInfo *SPDRInfo, pdrContext *PDRCreationContext) error {
+	if outerHeaderRemoval, err := pdr.OuterHeaderRemovalDescription(); err == nil {
+		spdrInfo.PdrInfo.OuterHeaderRemoval = outerHeaderRemoval
+	}
+	if farid, err := pdr.FARID(); err == nil {
+		spdrInfo.PdrInfo.FarId = pdrContext.getFARID(farid)
+	}
+	if qerid, err := pdr.QERID(); err == nil {
+		spdrInfo.PdrInfo.QerId = pdrContext.getQERID(qerid)
+	}
+
+	pdi, err := pdr.PDI()
+	if err != nil {
+		return fmt.Errorf("PDI IE is missing")
+	}
+
+	if sdfFilter, err := pdr.SDFFilter(); err == nil {
+		if sdfFilterParsed, err := ParseSdfFilter(sdfFilter.FlowDescription); err == nil {
+			spdrInfo.PdrInfo.SdfFilter = &sdfFilterParsed
+			// log.Printf("Sdf Filter Parsed: %+v", sdfFilterParsed)
+		} else {
+			return err
+		}
+	}
+
+	//Bug in go-pfcp:
+	//if fteid, err := pdr.FTEID(); err == nil {
+	if teidPdiId := findIEindex(pdi, 21); teidPdiId != -1 { // IE Type F-TEID
+		if fteid, err := pdi[teidPdiId].FTEID(); err == nil {
+			var teid = fteid.TEID
+			if pdrContext.ResourceManager.FTEIDM != nil {
+				if fteid.HasCh() {
+					var allocate = true
+					if fteid.HasChID() {
+						if teidFromCache, ok := pdrContext.TEIDCache[fteid.ChooseID]; ok {
```

Comment:
Try to hide teid cache usage in pdrContext. This function shouldn't know anything about implementation details including cache